src/snowflake/cli/plugins/stage/md5.py

```diff
@@ -126,21 +127,29 @@ def file_matches_md5sum(local_file: Path, remote_md5: str | None) -> bool:
         (_, num_chunks) = md5_and_chunks
         file_size = os.path.getsize(local_file)
 
-        # chunk size may not be a clean multiple of EST_CHUNK_GRANULARITY_BYTES
-        # try this first then fall back to the usual detection method
-        # at time of writing this would trigger around 80GiB for python connector
+        # If this file uses the maximum number of parts supported by the cloud backend,
+        # the chunk size is likely not a clean multiple of a megabyte. Try reverse engineering
+        # from the file size first, then fall back to the usual detection method.
+        # At time of writing this logic would trigger for files >= 80GiB (python connector)
         if num_chunks == S3_MAX_PARTS:
             chunk_size = max(math.ceil(file_size / S3_MAX_PARTS), S3_MIN_PART_SIZE)
             if compute_md5sum(local_file, chunk_size) == remote_md5:
                 return True
 
-        # try some common chunk size granularities
-        # smaller values create more even chunks; e.g. 13MB => [8MB, 6MB]
-        for granularity in [S3_CHUNK_SIZE, 1024**2]:
-            granules = 1 + ((file_size - 1) // (num_chunks * granularity))
-            chunk_size = granules * granularity
-            md5 = compute_md5sum(local_file, chunk_size)
-            if md5 == remote_md5:
+        # Estimates the chunk size the multi-part file must have been uploaded with
+        # by trying chunk sizes that give the most evenly-sized chunks.
+        #
+        # First we'll try the chunk size that's a multiple of S3_CHUNK_SIZE (8mb) from
+        # the python connector that results in num_chunks, then we'll do the same with
+        # a smaller granularity (1mb) that is used by default in some AWS multi-part
+        # upload implementations.
+        #
+        # We're working backwards from num_chunks here becuase it's the only value we know.
+        for chunk_size_alignment in [S3_CHUNK_SIZE, ONE_MEGABYTE]:
+            # +1/-1 ensures we don't add an additional chunk if exactly chunk-aligned size
```

Comment:
```suggestion
            # +1 because we need at least one chunk when file_size < num_chunks * chunk_size_alignment
            # -1 because we don't want to add an extra chunk when file_size is an exact multiple of num_chunks * chunk_size_alignment
```